feat: Add vector module

Otard95 opened the pull request:
Just wanted to be able to work with vectors as well.
There is potential for improvements though. The errors ar not displayed in the best way and the `__tostring` could be used for display the vec as virutal text.

But at least its functional and you can read the values like so
```lua
a = vec(1,2,3)
x = vec[1]
y = vec[2]
z = vec[3]
```

Otard95 committed:
feat: Add vector module

jbyuki commented:
Looks very nice! 
I think the printing is missing. 
You would need to hook up a section reference here:

https://github.com/jbyuki/quickmath.nvim/blob/4c2e29c82565e3a9847bbf7b7ff06c30c6bf2dfe/src/quickmath.lua.t#L83-L86

Then define a section similar to this:

https://github.com/jbyuki/quickmath.nvim/blob/4c2e29c82565e3a9847bbf7b7ff06c30c6bf2dfe/src/quickmath.lua.t#L98-L103

Also some documentation in the README would be nice for anyone who would want to use this without having to introspect the code.

Otard95 commented:
I'll see if I can figure out the printing.
But documentation I can do for sure!

jbyuki commented on lua/quickmath/init.lua:
```diff
@@ -116,6 +120,227 @@ function go_eol()
 
 end
 
+function vec(...)
```
Just a small comment, the syntax for vector initialization with `a = vec {1,2,3}` instead of `a = vec(1,2,3)` could be nicer. I'll let you decide because it's strictly design and it's "your" code.

jbyuki commented on lua/quickmath/init.lua:
```diff
@@ -116,6 +120,227 @@ function go_eol()
 
 end
 
+function vec(...)
```
Just a small comment, the syntax for vector initialization with `a = vec {1,2,3}` instead of `a = vec(1,2,3)` could be nicer. I'll let you decide because it's strictly design and it's "your" code.

Otard95 replied:
Not sure I'm familiar with that type of initialization. How would the signature look for that?

jbyuki replied:
It's a small trick in lua. If the function takes a single argument, like a table, the parenthesis for the call can be omitted.

```lua
function myfun(a)
   -- a is a table
   -- do something with a
end

-- normally you would call the function like this
myfun({1,2,3,4})

-- but lua allows this
myfun {1,2,3,4}
```

Works also for strings.

Otard95 replied:
Ok I'll test it out!

Otard95 replied:
I opted for allowing all variants.

jbyuki commented on lua/quickmath/init.lua:
```diff
@@ -116,6 +120,227 @@ function go_eol()
+    -- Normalize the vector
+    local mag = self:mag()
+    if mag == 0 then
+        return self
```
Might be better to clone here.
Otherwise it could lead to unexpected results because the vector is returned as reference.

jbyuki replied:
Actually ignore this because norm seem to work in-place.
But other places might be subject to this.

Otard95 replied:
Some places its nice to mutate and others clone. But maybe its better to be consistent and say that vectors are immutable. The main difference would be:
```
-- mutable
a = vec(1,2,3)
a:norm()
-- immutable
a = vec(1,2,3)
a = a:norm()
```

in either case you'll be able to chain functions
```
a = vec(1,2,3)
a:scale(5):limit(10)
```

jbyuki replied:
Immutable could be nice. It makes sense with how quickmatch works. The result in quickmath is only printed if assigned, with immutability, it has be assigned to do anything.

Otard95 replied:
Agreed. I'll make the change!

jbyuki commented on lua/quickmath/init.lua:
```diff
@@ -116,6 +120,227 @@ function go_eol()
+    for i = 1, self.dim do
+        self[i] = self[i] / mag
+    end
+    return self
```
Same here and other similiar places.

jbyuki replied:
Ignore this.

Otard95 committed:
feat: Make vectors immutable

By doing so the set functions is exactly the same as creating a new vec

Otard95 committed:
docs: Added documentation

Otard95 commented:
Made the vectors immutable and removed the `set()` functions as that now becomes the same as creating a new vec.
Also removed the `__eq()` as its not really relevant for this.

To see the docs rendered properly you can use [this link](https://github.com/Otard95/quickmath.nvim/tree/feat/vector)

Otard95 committed:
feat: Allow for vector init like `vec {1,2,3}`

Otard95 committed:
chore: Fix indenting

Otard95 committed:
feat: Add virtual text printing for vectors

Otard95 commented:
> I think the printing is missing.

Got the printing working now too!

jbyuki commented:
Awesome 🚀 
I will do a quick check, but I feel like it should be good to merge.

jbyuki commented:
Everything seem to work perfectly.
Documentation looks also very good and complete.
Thank you @Otard95
